### ISSUE TITLE
Add option to disable .detach() in MuQMuLan forward pass to preserve autograd graph

### ISSUE THREAD

**Issue by mcrespoae** [tencent-ailab/MuQ#7]
Hi! Thanks for the great work on MuQMulan.

I’m working on a project where I need to optimize some input perturbations using gradient-based methods. However, it seems like the model's output embeddings are detached from the graph, which prevents gradients from flowing back to the inputs.

Would you consider adding an option to keep the computation graph intact (e.g., a detach=False flag), or is there a reason for forcing .detach() in the forward pass?

Thanks in advance!

**Comment by juhayna-zh** [comment]
Hi! Can you indicate which specific line of code (or file) contains the unexpected `detach`? This may help me better identify the issue.

**Comment by mcrespoae** [comment]
Hi, thanks for the quick response! Here's a minimal reproducible example that seems to demonstrate the issue (it looks like the gradients do not flow back from the model's output, even though `requires_grad=True`:

```python 
    mulan_model = MuQMuLan.from_pretrained("OpenMuQ/MuQ-MuLan-large").to(DEVICE)
    dummy_audio = torch.randn(1, 16000).to(DEVICE).requires_grad_()

    emb_clean = mulan_model(wavs=dummy_audio, parallel_processing=True)
    print("emb_clean.grad_fn:", emb_clean.grad_fn)  # Should not be None if attached to graph which is correct

    dummy_audio_delta = torch.randn(1, 16000).to(DEVICE).requires_grad_()
    optimizer = torch.optim.Adam([dummy_audio_delta], lr=1e-2, weight_decay=1e-3, maximize=True)
    dummy_audio_total = dummy_audio + dummy_audio_delta
    emb_modified = mulan_model(wavs=dummy_audio_total, parallel_processing=True)
    loss = nn.functional.mse_loss(emb_clean, emb_modified)
    print("emb_modified.requires_grad:", emb_modified.requires_grad)  # This returns True
    print("emb_modified.grad_fn:", emb_modified.grad_fn)  # Should not be None if attached to graph which is correct
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    print("dummy_audio_delta.grad:", dummy_audio_delta.grad)  # This returns None
```

As you can see, `dummy_audio_delta.grad` is `None`, which suggests that the model’s output is detached from the computation graph (does the model use @torch.no_grad() internally or return .detach() explicitly?). 

Let me know if I’m missing something here or if this behavior is expected.

Thanks!

**Comment by juhayna-zh** [comment]
Hey, I took some time to understand your question, but luckily I found the key point.  

**First**, I recommend using `mulan_model.mulan_module.get_audio_latents(audio)` instead of `mulan_model(wavs=audio, parallel_processing=True)` in your case. This is because directly calling the `forward` method of `mulan_model` performs some preprocessing operations, such as splitting the audio into 10-second segments to match MuQ-MuLan's input. Using the `get_audio_latents` method, however, skips these preprocessing steps, which helps preserve the gradients of the input tensor.  

**Second**,  
In the file `src/muq/muq/models/muq_model.py`, you can find four `@torch.no_grad()`. You can simply remove these lines of code, as they only involve preprocessing of the input data. I don’t think removing them will cause any side effects, but since this is a relatively rare requirement, I won’t make the changes for now unless more people report this issue later.  

For now, I suggest you manually `git clone` this repository, remove the `@torch.no_grad()` lines in `src/muq/muq/models/muq_model.py` locally, and then install your modified version of muq using `pip install -e .`. This should give you the correct results.  

**Lastly**, I might remind you that MuQ's audio input requires a sampling rate of 24,000 Hz. 

So here’s my recommended code:  

```python
mulan_model = MuQMuLan.from_pretrained("OpenMuQ/MuQ-MuLan-large").to(DEVICE)
dummy_audio = torch.randn(1, 24000*10).to(DEVICE).requires_grad_() # 24K Hz input, recommend to use 10 seconds

emb_clean = mulan_model.mulan_module.get_audio_latents(dummy_audio) # to replace `mulan_model(wavs=audio, parallel_processing=True)`
print("emb_clean.grad_fn:", emb_clean.grad_fn)  # Should not be None if attached to graph which is correct

dummy_audio_delta = torch.randn(1, 24000*10).to(DEVICE).requires_grad_()
optimizer = torch.optim.Adam([dummy_audio_delta], lr=1e-2, weight_decay=1e-3, maximize=True)
dummy_audio_total = dummy_audio + dummy_audio_delta
emb_modified = mulan_model.mulan_module.get_audio_latents(dummy_audio_total)
loss = nn.functional.mse_loss(emb_clean, emb_modified)
print("emb_modified.requires_grad:", emb_modified.requires_grad)  # This returns True
print("emb_modified.grad_fn:", emb_modified.grad_fn)  # Should not be None if attached to graph which is correct
optimizer.zero_grad()
loss.backward()
optimizer.step()
print("dummy_audio_delta.grad:", dummy_audio_delta.grad)  # This returns correct result, after modify the code in `src/muq/muq/models/muq_model.py` 
```  

Wishing you smooth progress in your project :)
